extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/VariableWidthTransfer.java

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright (c) 2016-2023 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.parquet.table.transfer;
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.chunk.ObjectChunk;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.engine.rowset.RowSequence;
+import io.deephaven.engine.table.ChunkSource;
+import io.deephaven.engine.table.ColumnSource;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * TODO Add comments
+ */
+abstract class VariableWidthTransfer<T, E, B> implements TransferObject<B> {
+    private ObjectChunk<T, Values> chunk;
+    protected B buffer; // Not final because we might need to resize it in case of overflow
+    private final ColumnSource<?> columnSource;
+    private final RowSequence.Iterator tableRowSetIt;
+    private final ChunkSource.GetContext context;
+    private final int maxValuesPerPage;
+    private final int targetPageSize;
+    private int currentChunkIdx;
+    /**
+     * Cached value which took us beyond the page size limit. We cache it to avoid re-encoding.
+     */
+    @Nullable
+    private EncodedData cachedValue;
+
+    VariableWidthTransfer(@NotNull final ColumnSource<?> columnSource, @NotNull final RowSequence tableRowSet,
+            final int maxValuesPerPage, final int targetPageSize, @NotNull final B buffer) {
+        this.columnSource = columnSource;
+        this.tableRowSetIt = tableRowSet.getRowSequenceIterator();
+        this.targetPageSize = targetPageSize;
+        Assert.gtZero(maxValuesPerPage, "targetPageSize");
+        this.maxValuesPerPage = maxValuesPerPage;
+        Assert.gtZero(maxValuesPerPage, "maxValuesPerPage");
+        this.context = columnSource.makeGetContext(maxValuesPerPage);
+        this.currentChunkIdx = 0;
+        this.buffer = buffer;
+    }
+
+    @Override
+    public final B getBuffer() {
+        return buffer;
+    }
+
+    final public boolean hasMoreDataToBuffer() {
+        // Unread data present either the table or in the chunk
+        return tableRowSetIt.hasMore() || chunk != null;
+    }
+
+    // TODO Add comments
+    class EncodedData {
+        E data;
+        int numValues; // TODO Add comments
+        int numBytes;
+
+        EncodedData(@NotNull final E data, final int numBytes) {
+            this.data = data;
+            this.numBytes = numBytes;
+            this.numValues = 1;
+        }
+
+        EncodedData(@NotNull final E data, final int numValues, final int numBytes) {
+            this(data, numBytes);
+            this.numValues = numValues;
+        }
+    }
+
+    // TODO Add comments about what to be done before and after calling this method
+    final void transferOnePageToBufferHelper() {
+        if (!hasMoreDataToBuffer()) {
+            return;
+        }
```

Comment:
If we're hitting this, maybe there's a bug in the calling code?

Reply:
Yup, you are right, I can remove this check